lib/harmony/resources/resources.go

```diff
@@ -0,0 +1,180 @@
+package resources
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"runtime"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/pbnjay/memory"
+	"github.com/samber/lo"
+	"golang.org/x/sys/unix"
+
+	ffi "github.com/filecoin-project/filecoin-ffi"
+
+	"github.com/filecoin-project/lotus/lib/harmony/harmonydb"
+	cl "github.com/filecoin-project/lotus/lib/harmony/resources/miniopencl"
+)
+
+var LOOKS_DEAD_TIMEOUT = 10 * time.Minute // Time w/o minute heartbeats
+
+type Resources struct {
+	Cpu       int
+	Gpu       float64
+	GpuRam    uint64
+	Ram       uint64
+	MachineID int
+}
+type Reg struct {
+	Resources
+	shutdown atomic.Bool
+}
+
+var logger = logging.Logger("harmonytask")
+
+var lotusRE = regexp.MustCompile("lotus-worker|lotus-harmony|yugabyted|yb-master|yb-tserver")
+
+func Register(db *harmonydb.DB, hostnameAndPort string) (*Reg, error) {
+	var reg Reg
+	var err error
+	reg.Resources, err = getResources()
+	if err != nil {
+		return nil, err
+	}
+	ctx := context.Background()
+	{ // Learn our owner_id while updating harmony_machines
+		var ownerID []int
+		err := db.Select(ctx, &ownerID, `SELECT id FROM harmony_machines WHERE host_and_port=$1`, hostnameAndPort)
+		if err != nil {
+			return nil, fmt.Errorf("could not read from harmony_machines: %w", err)
+		}
+		if len(ownerID) == 0 {
+			err = db.QueryRow(ctx, `INSERT INTO harmony_machines 
+		(host_and_port, cpu, ram, gpu, gpuram) VALUES
+		($1,$2,$3,$4,$5) RETURNING id`,
+				hostnameAndPort, reg.Cpu, reg.Ram, reg.Gpu, reg.GpuRam).Scan(&reg.Resources.MachineID)
+			if err != nil {
+				return nil, err
+			}
+
+		} else {
+			reg.MachineID = ownerID[0]
+			_, err := db.Exec(ctx, `UPDATE harmony_machines SET
+		   cpu=$1, ram=$2, gpu=$3, gpuram=$4 WHERE id=$6`,
+				reg.Cpu, reg.Ram, reg.Gpu, reg.GpuRam, reg.Resources.MachineID)
+			if err != nil {
+				return nil, err
+			}
+		}
+		CleanupMachines(context.Background(), db)
```

Comment:
This returns rows affected? If so, we should probably log that when it's more than 0